### DISCUSSION TITLE
[Feature Request] Realtime backups during gameplay

### DISCUSSION THREAD

**Discussion by PMTabs** [MikeMaximus/gbm#251] [Ideas]
An option to make backups during play when the 'date/time modified' updates would be great. I currently use FreeFileSync's realtime backup to monitor certain games whose savegames are more volatile, or in the case of one game, where I'm playing in Permadeath Mode but don't want to lose my progress unfairly because of a bug or physics glitch, which is usually what kills me!

If GBM could do this itself, I could ditch the FFS tool.

Cheers!

**Comment by Darthagnon** [comment]
Interesting idea, but scary for most games, as concurrent read and write could corrupt stuff. I understand how saving before permadeath to save-cheese would be useful, but usually, for the game to re-detect the previous saves, wouldn't you have to close and reload it? e.g. I play Vagante, a roguelike, and save-cheese it, and for it to detect a save from before I died, I have to recopy the save file and close and re-open the game. At which point, the current backup/restore method would be enough...

**Comment by MikeMaximus** [comment]
Thanks for the suggestion PMTabs.  

I've had this feature brought up a few times over the years, but I've always shot it down partly to what Darthagnon has mentioned about file corruption.  It would be a great feature for some games, but I'm not sure it could ever be 100% safe or reliable to automatically backup saved games during game play.

That being said it's not completely off the table, it's a feature I may add as an option eventually for those that want to risk using it for certain games.

I know you're using FreeFileSync right now, but the following steps are what I usually suggest for saved games that you want to backup or restore while a game is running.

1.  Disable monitoring for the game in the Game Manager.
![image](https://user-images.githubusercontent.com/15719732/136698681-9b5c30cd-6167-4f65-b720-702c8e2a9336.png)
2.  Alt-Tab to GBM when you're fairly sure it's safe to do a backup or restore, such as the main menu.
3.  Select the game from the list on the main GBM window, click the "Backup" or "Restore" button.
![image](https://user-images.githubusercontent.com/15719732/136698752-c8b7fb29-603d-4d1b-a06c-c05fcab922c9.png)
4.  Switch back to the game when the operation is complete.

I've used the above steps and it's worked just fine for some games, but there's also a lot of games where it didn't work properly.  It's always risky and will differ from game to game.

I'm going to move this feature request to the Discussions area, feel free to continue the discussion there!

**Comment by MikeMaximus** [comment]
In v1.3.6 or later, you can now use a [Global Hotkey](https://mikemaximus.github.io/gbm-web/manual.html#settings_globalhotkeys) or the Backup and Restore buttons on the main window to make saved game backups (or restore) during game play.

See the [Backup and Restore during Gameplay](https://mikemaximus.github.io/gbm-web/manual.html#livebackup) section in the manual for details on how this works.

**Remember that support for this will vary from game to game.  It's up to the user to determine at which point it's safe (if ever) to backup or restore saves while the game is running.**

**Comment by DcMityay** [comment]
Great, it would be nice to see an option to automatically create a backup after a certain amount of time after starting the game. For example set the interval every 5(10, 15, etc.) minutes
This would be useful in case of Blue Screen of Death, power failure and other cases...

**Reply by MikeMaximus** [reply]
Since this continues to be a popular request and the framework for it was pretty much done, I've decided to add this feature in the next release.

You will be able to set a backup time interval (minimum 1 minute) for each game in the Game Manager.  While the game is detected backups will run automatically on that interval.

This feature already available in the v137 branch if anyone is adventurous enough. 😁

**Reply by DcMityay** [reply]
Installed version 137, the files are automatically copied! Cool!
But there is a problem (((
For example, in the game BALDUR'S GATE 3, save files have a large volume (over 400 megabytes) and with frequent copying there is a problem with free space on the disk ((((
Is there a possibility to add a check, if the files since the last backup were not changed, then skip the creation of the current backup? To save space on the hard disk? for save multiple backups
thx

**Reply by MikeMaximus** [reply]
>Is there a possibility to add a check, if the files since the last backup were not changed, then skip the creation of the current backup? To save space on the hard disk? for save multiple backups thx

This has been requested a lot recently as well, mostly because of BG3's massive save files.  I'll look into it again to see how much work it'll be.

I've been suggesting to use the [Differential Backup](https://mikemaximus.github.io/gbm-web/manual.html#differential) option to save disk space with BG3, you should be able to use it in combination with the new time interval option.

**Reply by DcMityay** [reply]
Yeah, I already thought of that, but I wanted to have backups of earlier games as well. 
gotta try it ...

**Reply by DcMityay** [reply]
Is there an option to create a new differential backup for each new game launch?

**Reply by MikeMaximus** [reply]
> Is there an option to create a new differential backup for each new game launch?

No sorry there isn't an option for that.

You can use the "Save multiple backups" setting in combination with the "Differential backup" setting.  Once the limit you set is reached old backup sets will be deleted automatically as new ones get created.  You'll just have to find a reasonable compromise with how many old sets you want to keep at once and how often you want it to backup.
